dali/pipeline/operators/util/reshape.cc

```diff
@@ -0,0 +1,198 @@
+  DALI_ENFORCE(shape.sample_dim() == 1 || (shape.sample_dim() == 2 && shape.num_samples() == 1),
+    "Reshape: shape input must be a list of 1D tensors or a single 2D tensor");
+  if (shape.sample_dim() == 2) {
+    auto shape_tensor = shape[0];
+    int N = shape_tensor.shape[0];
+    int dim = shape_tensor.shape[1];
+    output_shape_.resize(N, dim);
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < dim; d++) {
+        output_shape_.tensor_shape_span(i)[d] = *shape_tensor(i, d);
+      }
+    }
+  } else {
+    int N = shape.num_samples();
+    int sample_dim;
+    for (int i = 0; i < N; i++) {
+      int current_sample_dim = shape.tensor_shape_span(i)[0];
+      if (i) {
+        DALI_ENFORCE(current_sample_dim == sample_dim,
+          "Reshape: all samples must have the same number of dimensions");
+      } else {
+        sample_dim = current_sample_dim;
+        output_shape_.resize(N, sample_dim);
+      }
+
+      for (int d = 0; d < sample_dim; d++) {
+        output_shape_.tensor_shape_span(i)[d] = shape.tensor_data(i)[d];
+      }
+    }
+  }
+}
+
+template <typename Backend>
+template <typename TensorListLike>
+void Reshape<Backend>::ShapeFromInput(const TensorListLike &tl) {
+  TYPE_SWITCH(tl.type().id(), type2id, type, (int32_t),
+    (this->ShapeFromInput(view<const type>(tl));),
+    (DALI_FAIL("Reshape: shape input must have integral type; got: " + tl.type().name());)
+  );  // NOLINT
+}
+
+template <typename Backend>
+void Reshape<Backend>::CalculateOutputShape(const Workspace &ws) {
+  input_shape_ = ws.template InputRef<Backend>(0).shape();
+  const int N = input_shape_.num_samples();
+  switch (shape_source_) {
+    case ShapeSource::Arg:
+      if (output_shape_.num_samples() != N) {
+        output_shape_ = kernels::uniform_list_shape(N, uniform_shape_);
+      }
+      break;
+    case ShapeSource::ArgInput:
+      ShapeFromInput(ws.ArgumentInput("shape"));
+      break;
+    case ShapeSource::Input:
+      ShapeFromInput(ws.template InputRef<CPUBackend>(1));
+      break;
+    case ShapeSource::None:
+      output_shape_ = input_shape_;
+      return;
+  }
+
+  DALI_ENFORCE(output_shape_.num_samples() == N,
+    "Reshape: the new shape mush have same number of samples. Got " +
+    to_string(output_shape_.num_samples()) + ", expected " + to_string(N));
+
+  for (int i = 0; i < N; i++) {
+    auto actual_volume = volume(input_shape_.tensor_shape_span(i));
+    auto requested_volume = volume(output_shape_.tensor_shape_span(i));
+    DALI_ENFORCE(actual_volume == requested_volume,
+      "Input and output samples should have the same number of elements");
+  }
+}
+
+template <typename Backend>
+TensorLayout Reshape<Backend>::GetOutputLayout(const Workspace &ws) const {
+  if (!layout_.empty()) {
+    DALI_ENFORCE(output_shape_.sample_dim() == layout_.ndim(), "Reshape: requested layout '" +
+      layout_.str() + "' not compatible with " + to_string(output_shape_.sample_dim()) + "D shape");
```

Comment:
what does `D` mean here?

Reply:
After building the string it is 1D, 2D, 3D, etc.